src/vm/diagnosticsprotocol.h

```diff
@@ -49,6 +56,394 @@ bool TryParseString(uint8_t *&bufferCursor, uint32_t &bufferLen, const T *&resul
+    template <typename T>
+    struct HasFlatten
+    {
+        using yes = uint16_t;
+        using no  = uint8_t;
+        template <typename U, U u> struct Has;
+        template <typename U> static constexpr yes& test(Has<bool (U::*)(void*), &U::Flatten>*);
+        template <typename U> static constexpr no& test(...);
+        static constexpr bool value = sizeof(test<T>(int())) == sizeof(yes);
+    };
+
+    // template meta-programming to check for uint16_t(GetSize)() member function
+    template <typename T>
+    struct HasGetSize
+    {
+        using yes = uint16_t;
+        using no  = uint8_t;
+        template <typename U, U u> struct Has;
+        template <typename U> static constexpr yes& test(Has<uint16_t(U::*)(), &U::GetSize>*);
+        template <typename U> static constexpr no& test(...);
+        static constexpr bool value = sizeof(test<T>(int())) == sizeof(yes);
+    };
+
+    // template meta-programming to check for const T*(TryParse)(BYTE*,uint16_t&) member function
+    template <typename T>
+    struct HasTryParse
+    {
+        using yes = uint16_t;
+        using no  = uint8_t;
+        template <typename U, U u> struct Has;
+        template <typename U> static constexpr yes& test(Has<const U* (U::*)(BYTE*, uint16_t&), &U::TryParse>*);
+        template <typename U> static constexpr no& test(...);
+        static constexpr bool value = sizeof(test<T>(int())) == sizeof(yes);
+    };
+
+    // Encodes the messages sent and received by the Diagnostics Server.
+    //
+    // Payloads that are fixed-size structs don't require any custom functionality.
+    //
+    // Payloads that are NOT fixed-size simply need to implement the following methods:
+    //  * uint16_t GetSize()                                     -> should return the flattened size of the payload
+    //  * bool Flatten(BYTE *lpBuffer)                           -> Should serialize and write the payload to the provided buffer
+    //  * const T *TryParse(BYTE *lpBuffer, uint16_t& bufferLen) -> should decode payload or return nullptr
+    class IpcMessage
+    {
+    public:
+        // Create an outgoing IpcMessage from a header and a payload
+        template <typename T>
+        IpcMessage(IpcHeader header, T& payload)
+            : m_Header(header), m_Size(0), m_pData(nullptr)
+        {
+            CONTRACTL
+            {
+                NOTHROW;
+                GC_TRIGGERS;
+                MODE_PREEMPTIVE;
+            }
+            CONTRACTL_END;
+
+            FlattenImpl<T>(payload);
+        };
+
+        // Create an incoming IpcMessage from an incoming buffer
+        IpcMessage(::IpcStream* pStream)
+            : m_Header(), m_Size(0), m_pData(nullptr)
+        {
+            CONTRACTL
+            {
+                NOTHROW;
+                GC_TRIGGERS;
+                MODE_PREEMPTIVE;
+            }
+            CONTRACTL_END;
+
+            TryParse(pStream);
```

Comment:
Suggest pulling this out and having the caller check for success

Reply:
I moved this logic into an `Initialization` function which pushes the onus of error checking to the user.